Reference Specification for API first approach (#93)
nodeshift/nodejs-reference-architecture: docs/development/apifirst.md

---
sidebar_position: 6
---

# REST APIs

Building [RESTFull](https://www.redhat.com/en/topics/api/what-is-a-rest-api) APIs is a typical use
case for Node.js. There are two typical approaches:

* API First - define the API, use tools to help scaffold and then fill in the implementation.

> * Code First - implemhdawson-patch-1ment the APIs and then document based on the API exposed.

Author: ```suggestion
* Code First - implement the APIs and then generate documentation based on exposed API
```

The team's experience is that the API first approach based on [OpenAPI](https://swagger.io/specification/)
provides benefits in both initial implementation and maintenance and our recommended packages and
guidance is based on that approach, particularly if one or more of the following are true:

- New project without existing API
- Client side and backend are developed by different teams that need way to communicate API changes
- Development on clients and backends starts around the same time giving developers ability to mock API based on OpenAPI spec

> > NOTE: This section of the reference arcahitecture focusses on building RESTfull APIs. 

Reviewer: ```suggestion
> NOTE: This section of the reference architecture focuses on building RESTfull APIs. 
```

This API contract acts as a central draft keeping all your team members aligned on what your API’s objectives are and how your API’s resources are exposed. The finalization of the contract allows the team to build the interface of the application.

After this, the cross-functional teams rely on this interface to build the rest of the application independent of each other. In practice, teams can generate backends stubs and fully functional client libraries to avoid deviation from specification set in the OpenAPI spec file.

## Code First vs API first

Code first approach provides libraries that understand server side backend structure and generate respective OpenAPI files. 
In this approach full control over API lies within server side team - generated OpenAPI file is read only and cannot be effectively
used to negotiate API between client and server.

API first approach uses OpenAPI file as source of truth. Both client and server side generate code based on the OpenAPI file.

## Recommended Packages

> List bellow provides comprehensive set of libraries that can be used for an end to end full stack application written in Node.JS, 

Reviewer 2: ```suggestion
List bellow provides comprehensive set of libraries that can be used for an end to end full stack application written in Node.js, 
```

Express.js as well as client side applications.

### Code Generation Tools

> [openapitools/openapi-generator-cli](https://www.npmjs.com/package/@openapitools/openapi-generator-cli)

Author: We do not have standardized format for packages. Some docs use < some use (

Reviewer 3: You mean in terms of the links?  If so we should probably choose one and try to be consistent. I always use []() but happy to go with what the overall consensus is.

The project is maintained by OpenAPI Generator Community and you can read the documentation
here:- <https://openapi-generator.tech>. It can be used as both a backend and client generator as follows:

**backend generator**

The nodejs-express-server option  can be used to generate Express.js based stub
implementations based on your OpenAPI file. 
```bash
npx openapi-generator-cli generate -g nodejs-express-server -i yourapi.json -o ./yourproject
```

**Client generator**

The typescript-node option can be used to generate a client for Node.js applications
that allows us to perform requests against another backend

```bash
openapi-generator generate -g typescript-node -i yourapi.json -o ./yourproject
```

### API mocking

[openapi-backend](https://www.npmjs.com/package/openapi-backend) allows you to mock based
on an OpenAPI definition by returning predefined strings. The library provides way not only
return predefined stubs but also perform validation or handle different use cases depending on request 

[@stoplightio/prism](https://www.npmjs.com/package/@stoplight/prism-http) allows you to

> automatically mock API using OpenAPI spec definitions. This package is recomended if you

Reviewer: ```suggestion
automatically mock API using OpenAPI spec definitions. This package is recommended if you
```

need is an out of the box way to mock API without any development involved.

### API validation middleware

[express-openapi-validator](https://www.npmjs.com/package/express-openapi-validator) is a validator
for express middleware that some of the build have used successfully.

> ### Creating/editing OpenAPI Specificaitons

Reviewer: ```suggestion
### Creating/editing OpenAPI Specifications
```

[swagger-editor](https://www.npmjs.com/package/swagger-editor) is the most popular editor

> which can be embeeded into existing server or run standalone. If you want to edit your

Reviewer: ```suggestion
which can be embedded into an existing server or run standalone. If you want to edit your
```

[vscode-openapi](https://github.com/42Crunch/vscode-openapi) is a VScode plugin for
building and validation of OpenAPI files that members of the team using vscode
have used successfully.

## Guidance

Based on the teams experience we recommend the following:

1. Define the API using OpenAPI 3.0, you can write the OpenAPI definitions in YAML or JSON formats, the team does not have a preference for one over the other.
2. Prefer generating code from OpenAPI file for both client and server. Generating code based on the specification will ensure that the same types, formats are used. This will enable your team to iterate on the specification without worry of getting out of sync.
3. When making changes in the OpenAPI file change it's [version](https://github.com/OAI/OpenAPI-Specification/blob/main/examples/v3.0/petstore-expanded.yaml#L3). Changed version will help others to detect what kind of changes were made.
4. When introducing breaking changes consider adding them as new endpoints by introducing another v2 prefix in the API path. 
5. Every path should have `operationId` field specified. This field is used by generator to generate method names for clients etc.

> 6. When building response objects preffer referencing predefined Objects and Enums in [Schemas](https://swagger.io/docs/specification/data-models/)

Reviewer: ```suggestion
6. When building response objects prefer referencing predefined Objects and Enums in [Schemas](https://swagger.io/docs/specification/data-models/)
```

---
sidebar_position: 6
---

# API First REST backends with OpenAPI Specification 

## Guidance (change guidance

> This reference describes what it means to use the API-first design approach with Node.js backends and JavaScript client applications. Reference is going to use OpenAPI standard along with supporting packages and tools we recomend to implement API first applications.

Reviewer 4: ```suggestion
This reference describes what it means to use the API-first design approach with Node.js backends and JavaScript client applications. Reference is going to use OpenAPI standard along with supporting packages and tools we recommend to implement API first applications.
```

> NOTE: This guide focused on the OpenAPI Specification for building RESTfull services. 
For GraphQL please refer to the dedicated [GraphQL Guide][]

### What "API-first" mean?

With the API-first approach, designing the API is the first priority before writing any code. Design of the API involves thorough thinking and planning through collaboration with different teams (both client and backend side). This result in high-level documentation describing the intent of the API and ability to build client without waiting for server to be finished.

This API contract acts as a central draft keeping all your team members aligned on what your API’s objectives are and how your API’s resources are exposed. The finalization of the contract allows the team to build the interface of the application.

After this, the cross-functional teams rely on this interface to build the rest of the application independent of each other. In practice teams can generate backends stubs and fully functional client libraries to avoid deviation from specification set in the OpenAPI spec file.

> ### When to use OpenAPI first and when to avoid

Reviewer 5: IMHO, Contract-First is **ALWAYS** the right answer. Code first leads to re-work or mis-fitting APIs to use cases. Additionally, there are so many other advantages to Contract-First (automated test, better documentation, faster feedback, etc...) that make a code-first approach more of a "vanity" decision instead of a "technical" decision.

Author: Agree. Got suggestion on call to put out exactly that into details

#### @openapitools/openapi-generator-cli - <https://openapi-generator.tech>

CLI provides support for generating source code based on the OpenAPI spec. CLI supports wide range of languages and frameworks:

https://openapi-generator.tech/docs/generators

CLI has widespread usage across industry including many community projects at Red Hat.

Project is maintained by OpenAPI Generator Community

### Node.js backend generator

#### nodejs-express-server - <https://openapi-generator.tech/docs/generators/nodejs-express-server>

This generator generates Express.js based stub implementation based on your OpenAPI file. 
Generation can be done by using openapi-generator cli:

```bash
openapi-generator generate -g nodejs-express-server -i yourapi.json -o ./yourproject
```

### Client Side generator

#### typescript-node - <https://openapi-generator.tech/docs/generators/typescript-node>

This generator generates client for Node.js backend that allows us to perform requests against another backend
against a

```bash
openapi-generator generate -g typescript-node -i yourapi.json -o ./yourproject
```

### API mocking

#### openapi-backend - <https://www.npmjs.com/package/openapi-backend>

> 

Reviewer 5: @stoplightio/prism is my go-to for API mocking. It makes more in-depth use of the formats/types/constraints in it's mocking and requires no coding. For example in other tools when I specify a field like

```
myField:
    type: integer
    format: int64
    minimum: 1000
    maximum: 9999
```

Stoplight's `prism` will actually mock values which comply with the constraints of `minimum` and `maximum`.

Author: Great suggestion! Thank you

I would replace current suggestion with prism. We had explored both out of the box mocks and mocks that require handlers and found handler ones to be more flexible and working with every api without modification.

Prism is also more mature

> ---

Reviewer 3: I'd suggest we change the file to be apis.md instead of apifirst.md.  I'm thinking this section should talk about experience/recommendations for creating APIs with API first being the approach many teams have had success with/would recommend.

Author: Would that be confusing?  Many people will see it as "build your endpoints using express guide". 
Where this is dedicated to Patterns on Building restful API.

I intentionally avoided being biased in the doc by discussing topics in context of API first - meaning that team will need to understand it and want to use it.
